---
output: html_document
title: "Simulations for iPGS proof-of-concept study"
---

```{r setup, include=F}
knitr::opts_chunk$set(echo = F, message = F, warning = F, 
                      dpi = 150, fig.path = "../output/simulations/")
suppressMessages(silent <- lapply(
  c("knitr", "kableExtra", "tidyverse", "patchwork"), 
  library, character.only = TRUE))
theme_set(theme_bw())
```

# Simulation setup

Initial simulations will proceed as follows:

1. Simulate a single set of genotypes (N = 10,000 individuals, M = 100 variants) with a given MAF spectrum:
  * $MAF_m \sim uniform(0.01, 0.5)$
  * $G_m \sim binomial(2, MAF_m)$ for each variant $m$

Then, for each scenario (details below)...

2. Simulate one environment E and P phenotypes Y (1000 for type I error simulations, 200 for power simulations) with effect sizes based on overall specified variance explained:
  * $E \sim N(0, 1)$
  * $beta_E \sim sqrt(Var_{E,tot})$
  * $beta_{G,m} \sim N(0, sqrt(Var_{G,tot} / M))$
  * $beta_{GxE,m} \sim N(0, sqrt(Var_{GxE,tot} / M))$
  * $Y_i = \beta_EE_i + \beta_G^T\mathbf{G_i} + \beta_{GxE}^TE_i\mathbf{G_i} + \epsilon_i$, where $\epsilon_i \sim N(0, sqrt(1 - Var_{E,tot} - Var_{G,tot} - Var_{GxE,tot}))$

Then, for each combination of scenario and phenotype...

3. Test associations for each variant using interaction (genome-wide interaction study [GWIS]) and main effect (genome-wide association study [GWAS]) models, in the training set (70%):
  * $Y_i \sim \beta_0 + \beta_EE_i + \beta_{G}G_i + \beta_{GxE}G_iE_i$, with $\beta_{GxE}$ being the effect of interest
  * $Y_i \sim \beta_0 + \beta_{G}G_i$, with $\beta_{G}$ being the effect of interest
4. Get PGS weights:
  * For now, this will use a trivial strategy of simply pulling variant-specific effect sizes from the association tests without shrinkage.
  * $\beta_{iPGS,m} = \beta_{GWIS,m}$
  * $\beta_{PGS,m} = \beta_{GWAS,m}$
5. Calculate PGS:
  * $iPGS_i = \mathbf{G_i^T}\mathbf{\beta_{iPGS}}$
  * $PGS_i = \mathbf{G_i^T}\mathbf{\beta_{PGS}}$
6. Test PGS in the testing set (30%):
  * $Y_i \sim \beta_0 + \beta_EE_i + \beta_{iPGS}iPGS_i + \beta_{iPGSxE}iPGS_iE_i$
  * $Y_i \sim \beta_0 + \beta_EE_i + \beta_{PGS}PGS_i + \beta_{PGSxE}PGS_iE_i$
  
Variables to tune:

* E main effect (total variance)
  - Between 0 and 10%
* G main effect (total variance)
  - Between 0 and 10%
* GxE effect (total variance)
  - Between 0 and 5%
* Distribution of Y
* Measurement error in E
  - ICC: {50%, 75%, 100%}
* PRS calculation strategy
  - PGS
  - iPGS
  - rPGS
  - vPGS

```{r setup-sims}
power_scenario_df <- expand_grid(
  e_var_tot = c(0, 0.1),
  g_var_tot = c(0, 0.1),
  gxe_var_tot = seq(0, 0.2, 0.05),
  prob_nonzero = 1,
  n_sim = 200
) %>%
  mutate(tag = paste0("e", e_var_tot, "_g", g_var_tot, "_gxe", gxe_var_tot))

power_scenario_df %>%
  write_csv("../data/processed/simulations/power_scenarios.csv")
write(power_scenario_df$tags, "../data/processed/simulations/power_scenario_tags.txt")

t1e_scenario_df <- expand_grid(
  e_var_tot = 0.1,
  g_var_tot = c(0, 0.1),
  gxe_var_tot = 0,
  prob_nonzero = 1,
  n_sim = 1000
) %>%
  mutate(tag = paste0("e", e_var_tot, "_g", g_var_tot, "_gxe", gxe_var_tot))

t1e_scenario_df %>%
  write_csv("../data/processed/simulations/t1e_scenarios.csv")
write(t1e_scenario_df$tags, "../data/processed/simulations/t1e_scenario_tags.txt")


# tags <- scenario_df$tag
```

```{r summarize-sims}
calc_summaries <- function(tag) {
  res_fn <- paste0("../data/processed/simulations/", tag, "/test_res.csv")
  read_csv(res_fn) %>%
    group_by(pgs_type) %>%
    summarise(n_sig = sum(p.value < 0.05),
              prop_sig = n_sig / n(),
              mean_effect = mean(estimate),
              .groups = "drop")
}

t1e_scenario_res_df <- t1e_scenario_df %>%
  rowwise() %>%
  mutate(summ = list(calc_summaries(tag))) %>%
  unnest(summ) %>%
  mutate(pgs_type = factor(pgs_type, levels = c("ipgs", "pgs"), 
                           labels = c("iPGS", "PGS")))

power_scenario_res_df <- power_scenario_df %>%
  rowwise() %>%
  mutate(summ = list(calc_summaries(tag))) %>%
  unnest(summ) %>%
  mutate(pgs_type = factor(pgs_type, levels = c("ipgs", "pgs"), 
                           labels = c("iPGS", "PGS")),
         main_effects = gsub("_gxe.*", "", tag))
```

# Type I error

```{r t1e-summary}
t1e_scenario_res_df %>%
  mutate(g_var_tot = factor(g_var_tot)) %>%
  ggplot(aes(x = g_var_tot, y = prop_sig)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0.05, linetype = "dashed", color = "gray") +
  facet_wrap(vars(pgs_type), nrow = 1) +
  coord_cartesian(ylim = c(0, NA)) +
  labs(x = expression(Var["G,tot"]),
       y = "False positive rate")
```

# Power

```{r power-summary}
power_scenario_res_df %>%
  ggplot(aes(x = gxe_var_tot, y = prop_sig, 
             group = pgs_type, color = pgs_type)) +
  geom_point() +
  geom_line() + 
  facet_wrap(vars(e_var_tot, g_var_tot), labeller = "label_both") +
  scale_color_discrete(name = "PGS type") +
  coord_cartesian(ylim = c(0, NA)) +
  labs(x = expression(Var["GxE,tot"]),
       y = "Statistical power")
```

# Archive

```{r run-sims, eval=F}
source("simulations/simulation_funcs.R")

simulate_genotypes(N = 10000, M = 100, min_maf = 0.01, max_maf = 0.05)
g_mat <- readRDS("../data/processed/simulations/g_mat.rds")

simulate_phenotypes(scenario_df, g_mat, train_prop = 0.7)

for (tag in tags) test_associations(tag, g_mat, n_cores = 6)

for (tag in tags) get_pgs_weights(tag)

for (tag in tags) calculate_pgs(tag, g_mat)

for (tag in tags) test_pgs(tag, g_mat)
```

```{r first-sim-pipeline, eval=F}
set.seed(1)
n <- 10000  # Number of samples
m <- 100  # Number of genotypes
beta_gxe_prob_nonzero <- 0.5  # Probability of non-zero genotype effect

# Generate genotype matrix

generate_g_mat <- function(m, min_maf = 0.01, max_maf = 0.5) {
  mafs <- runif(m, min_maf, max_maf)
  sapply(mafs, function(maf) {
    rbinom(n, 2, maf)
  }, simplify = TRUE)
}

g_mat <- generate_g_mat(m)

# Generate genotype effect vector

true_beta_gxe_vec <- ifelse(runif(m) < beta_gxe_prob_nonzero, rnorm(m, 0, 1), 0)

# Simulate exposure and outcome

sim_df <- tibble(
  id = seq(1, n),
  e = rnorm(n, 0, 1),
  y = rnorm(n, (g_mat * sim_df$e) %*% true_beta_gxe_vec, 1)
)

# Split dataset

split_dataset <- function(n_total, train_prop) {
  train_idx <- sample(seq(1, n_total), round(n_total * train_prop), replace = FALSE)
  train_idx
}

train_idx <- split_dataset(n, 0.7)
test_idx <- setdiff(seq(1, n), train_idx)

# Estimate interaction betas for all genotypes

test_g_by_e <- function(g, e, y, df, covars = NULL) {
  regression_df <- df %>%
    rename(g = {{ g }}, e = {{ e }}, y = {{ y }})
  lm_form_str <- "y ~ g * e"
  if (!is.null(covars)) {
    lm_form_str <- paste0(lm_form_str, " + ", paste(covars, collapse = " + "))
  }
  lm(as.formula(lm_form_str), data = regression_df) %>%
    broom::tidy() %>%
    filter(term == "g:e")
}

gwis_res_df <- map_dfr(seq(1, m), function(g_idx) {
  sim_df$g <- g_mat[, g_idx]
  test_g_by_e("g", "e", "y", sim_df[train_idx, ])
})

# Calculate iPGS

calculate_pgs <- function(g_mat, beta_vec) {
  g_mat %*% beta_vec
}

sim_df$ipgs <- calculate_pgs(g_mat, gwis_res_df$estimate)

# Test PGS

test_pgs_by_e <- function(ipgs, e, y, df, covars = NULL) {
  regression_df <- df %>%
    rename(ipgs = {{ ipgs }}, e = {{ e }}, y = {{ y }})
  lm_form_str <- "y ~ ipgs * e"
  if (!is.null(covars)) {
    lm_form_str <- paste0(lm_form_str, " + ", paste(covars, collapse = " + "))
  }
  lm(as.formula(lm_form_str), data = df) %>%
    broom::tidy() %>%
    filter(term == "ipgs:e")
}

test_pgs_by_e("ipgs", "e", "y", sim_df[test_idx, ])
```

